### PR TITLE
Implement Planner support for Union ALL 

### PR DESCRIPTION
**Pull request by mxm**
The Union ALL planner support uses a `Union` operator in the `LogicalPlanner`
which generates a `UnionPlan`. The plan consists of a sub plan for each side of
the union. When the union is planned, both sides are executed as a branch of the
execution tree. The result is fed to the `MergePhase` which receives data from
both sub plans and merges them into buckets encoded by the node id and the input
id set during planning.

In `MergePhase` the number of different inputs was added to ensure that a merge
with multiple inputs can be detected. The input id is encoded at the beginning
of the 32bit bucket id.

### CODE REVIEW

**Comment by mxm** [comment]
@matriv @mfussenegger Ready for an initial review.

**Comment by mfussenegger** [comment]
tests are failing?

**Comment by mxm** [comment]
They were passing before locally and here in the PR. I only squashed the commits. Still passing locally. Need to investigate.

**Comment by mxm** [comment]
Now passing. Still investigating the cause of the previous test fails but maybe you could take a look at the PR in the meantime?

File: sql/src/main/java/io/crate/planner/node/dql/MergePhase.java [hunk 1]
```diff
@@ -60,16 +62,31 @@
...
+     * @param jobId The JobID of the entire execution.
+     * @param executionNodeId A unique execution id for this phase.
+     * @param name The name of the MergePhase.
+     * @param numUpstreams The number of nodes to expect data from.
+     * @param numInputs The number of different inputs to read data from.
+     * @param executionNodes The nodes where this MergePhase executes.
+     * @param inputTypes The types of the input rows.
+     * @param projections The projections to apply when merging.
+     * @param distributionInfo The default strategy to use when distributing the results of the MergePhase.
+     * @param positionalOrderBy The order by positions to apply if ordering is supplied.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: Maybe rephrase this to something like `The order by positions the input is pre-sorted on; setting this will result in a sorted merge`

File: sql/src/main/java/io/crate/planner/node/dql/MergePhase.java [hunk 2]
```diff
@@ -60,16 +62,31 @@
      */
     @Nullable private PositionalOrderBy positionalOrderBy;
 
+    /**
+     * Creates a MergePhase to combine the results from multiple node operations.
+     * @param jobId The JobID of the entire execution.
+     * @param executionNodeId A unique execution id for this phase.
+     * @param name The name of the MergePhase.
+     * @param numUpstreams The number of nodes to expect data from.
+     * @param numInputs The number of different inputs to read data from.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: Maybe clarify that this is the number of upstream phases? "different inputs" is a bit vague

File: sql/src/main/java/io/crate/planner/operators/Union.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            2,
+            Collections.singletonList(plannerContext.handlerNode()),
+            leftResultDesc.streamOutputs(),
+            Collections.emptyList(),
+            DistributionInfo.DEFAULT_BROADCAST,
+            null
+        );
+
+
+        return new UnionPlan(
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/operators/Union.java]: Hm - what we haven't considered here is that if we do a insert from subquery, we're unnecessarily moving all data to the handler before creating the insert requests - it would be possible to keep the data distributed and insert from each node directly (as long as there is no limit involved)

I guess this is okay for the initial support - but we'll have to consider how we can adapt the LogicalPlan building to support such cases. One option would be to make the `ensureOnHandler` logic part of the `LogicalPlan` - so that we can encapsulate different behaviours depending on the operator.

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+@ESIntegTestCase.ClusterScope(minNumDataNodes = 2, maxNumDataNodes = 5)
+@UseJdbc
+public class UnionIntegrationTest extends SQLTransportIntegrationTest {
+
+    private boolean setupDone;
+
+    @Before
+    public void beforeTest() {
+        if (!setupDone) {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: I think this doesn't work as you'd expect it to? setupDone is re-initialized per test. I'd have to be static and then you've to be careful with synchronization.

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+import io.crate.testing.UseJdbc;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.Matchers.arrayContaining;
+import static org.hamcrest.Matchers.arrayContainingInAnyOrder;
+import static org.hamcrest.Matchers.is;
+
+@ESIntegTestCase.ClusterScope(minNumDataNodes = 2, maxNumDataNodes = 5)
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: Usually we try not to restrict this unless there is a good reason to. I think it would be good to also have this test run with only 1 data node occasionally - a lot of our planner logic is different on 1 node vs many nodes.

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java (hunk 4 above)
Comment (mxm) [reply sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: Wasn't aware that JUnit creates a new instance of the class for every test.

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java (hunk 5 above)
Comment (mxm) [reply sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: Will make this from 1 to 5 then.

Comment (matriv) [reply sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: Should we do a class that extends it with minNumDataNodes=1 like `JoinSingleNodeIntegrationTest`

File: CHANGES.txt [hunk 6]
```diff
@@ -39,6 +39,11 @@ Breaking Changes
 Changes
 =======
 
+- Introduce support for UNION ALL to produce the combined result of two
+  queries, e.g.:
+
+    select * from t1 UNION ALL select * from t2
```
Comment (chaudum) [inline CHANGES.txt]: please capitalize keywords:
`SELECT * FROM t1 UNION ALL SELECT * FROM t2`

to make the example more explicit you could also explicitly say:
```
SELECT id FROM t1
UNION ALL
SELECT id FROM t2;
```

File: CHANGES.txt [hunk 7]
```diff
@@ -39,6 +39,11 @@ Breaking Changes
 Changes
 =======
 
+- Introduce support for UNION ALL to produce the combined result of two
```
Comment (chaudum) [inline CHANGES.txt]: ... of two or more queries.

**Comment by mxm** [comment]
The previous issues visible in some tests runs have been resolved.

File: sql/src/main/java/io/crate/planner/operators/Union.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/planner/operators/Union.java]: >Hm - what we haven't considered here is that if we do a insert from subquery, we're unnecessarily moving all data to the handler before creating the insert requests 

This is a general limitation of insert from subqueries because of https://github.com/mxm/crate/blob/maser/sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java#L92 , right?

```java
Insert into t values (id)
select id from t2
union all
select id from t3
```

>One option would be to make the ensureOnHandler logic part of the LogicalPlan - so that we can encapsulate different behaviours depending on the operator.

Seems like the best way to let the plan decide whether it needs a final merge.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/planner/operators/Union.java]: > This is a general limitation of insert from subqueries because of https://github.com/mxm/crate/blob/maser/sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java#L92 , right?

The `ensureOnHandler` call happens after the projection has been added. So no, this is not a general limitation.


> Seems like the best way to let the plan decide whether it needs a final merge.

So far we always tried to keep the executionPlan returned on `build`
distributed if possible. The *parent* operators were always responsible to add
the merge if needed. The problem is that with union we currently can't
represent it in a distribution way and let the parents decide if a merge is
required. This will need some interface extension / remodelling.

But as I said - I think this is out of scope of the work here. But we've to think about how we do this later on.

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java [hunk 8]
```diff
@@ -51,10 +56,10 @@
     private final Streamer<?>[] streamers;
     private final RamAccountingContext ramAccountingContext;
     private final int numBuckets;
-    private final BitSet exhausted;
+    private final Set<Integer> exhausted;
     private final PagingIterator<Integer, Row> pagingIterator;
-    private final IntObjectHashMap<PageResultListener> listenersByBucketIdx;
-    private final IntObjectHashMap<Bucket> bucketsByIdx;
+    private final Map<Integer, PageResultListener> listenersByBucketIdx;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: Is the IntObjectHashMap -> Map change necessary? If I understand the problem correctly only the BitSet was problematic?
If the Map change is unrelated please revert it. It just hides the real fix and there is nothing wrong with the hppc map, quite the opposite, it avoids integer boxing in this case.

**Comment by mfussenegger** [comment]
btw we might need some tests for union with array columns. Afaik currently we only ever load the values for array column using a fetchPhase - fetchPhase does source lookups; But since now we don't have the fetch support yet it will try to load the array values without source lookup - and I think that's not yet implemented in the `LuceneReferenceResolver`

File: sql/src/main/java/io/crate/planner/operators/Union.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/planner/operators/Union.java]: I think this whole logic of when a merge is applied or not, probably needs to be reworked in the future. As you said, the parents have to check but it would be nice if the children already decided if a merge is necessary for the following operations.

>The problem is that with union we currently can't
represent it in a distribution way and let the parents decide if a merge is
required.

I think I don't understand. The `Union` currently decides if a merge is required.

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java (hunk 8 above)
Comment (mxm) [reply sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: I thought you would comment on this :)

> Is the IntObjectHashMap -> Map change necessary? 

Yes, because the iterator of `IntObjectHashMap` doesn't have a `remove()` method which we need to iterate and remove values. Granted, we could remove later on but that would require to store a temporary list of to-be-removed-items.

Strictly speaking, `bucketsByIdx` wouldn't have to be converted, but I rather use once consistent interface for storing the bucket ids, then two different interfaces and data structures.

> it avoids integer boxing in this case.

Not in all cases, since we have also callbacks with `Integer` into the hppc map. Anyhow, this is an extreme micro-optimization.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: > I thought you would comment on this 

yep because it's very hard to distinguish between relevant and unrelevant changes, and we generally have a tendency to sneak in unrelated changes. That makes reviews only more difficult.

> Yes, because the iterator of IntObjectHashMap doesn't have a remove() 

I think `remove(idx)` could be called even while the iterator is used.

> Anyhow, this is an extreme micro-optimization. 

I'm well aware, I wanted to make a point that it's not *worse* than a HashMap.

Anyway, I'll leave this up to you

**Comment by mxm** [comment]
I've tested and verified that union with array and object columns work correctly. Adding a test case to ensure it stays that way.

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java (hunk 8 above)
Comment (mxm) [reply sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: >yep because it's very hard to distinguish between relevant and unrelevant changes, and we generally have a tendency to sneak in unrelated changes. That makes reviews only more difficult.

I'm really against sneaking in anything and always try to keep my commits clean. However, sometimes it is better to change a bit more and sacrifice clarity over minimal invasive changes.

>I think remove(idx) could be called even while the iterator is used.

Honestly, I had it like that in the beginning but I wasn't sure if that would introduce a bug. I think the resizing on remove could break the iterator.

>I'm well aware, I wanted to make a point that it's not worse than a HashMap.

I don't understand why you would prefer a non-standard API if it is not really beneficial. It just makes it harder for people to read the code.

**Comment by mxm** [comment]
@mfussenegger I've fixed the object collecting problem. Not sure if the fix is adequate directly in the LuceneRefenceResolver but I find that it is the best place to ensure source lookup conversion is performed: https://github.com/crate/crate/pull/6535/commits/b00f1e4d95e18c6a827084c86eacddec3aca044c

File: sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java [hunk 9]
```diff
@@ -67,6 +67,13 @@ public LuceneReferenceResolver(FieldTypeLookup fieldTypeLookup, IndexSettings in
     public LuceneCollectorExpression<?> getImplementation(Reference refInfo) {
         assert refInfo.granularity() == RowGranularity.DOC : "lucene collector expressions are required to be on DOC granularity";
 
+        if (refInfo.valueType().id() == ObjectType.ID && refInfo.granularity() == RowGranularity.DOC) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java]: There is an assertion for granularity == DOC just 2 lines above?

What about https://github.com/crate/crate/blob/b00f1e4d95e18c6a827084c86eacddec3aca044c/sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java#L133

why is that not working?

Comment (mxm) [reply sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java]: >There is an assertion for granularity == DOC just 2 lines above?

Yes, and?

The regular object lookup doesn't work because the reference has not been rewritten to a doc source lookup. This is done here:

https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/planner/operators/FetchOrEval.java#L336

>why is that not working?

Because the `Reference` points to `doc.table->obj` which is not an actual field in the field lookup. It needs to be `doc.table->_doc['obj']`.

**Comment by mxm** [comment]
I'd like to merge this to `union-new` within the next days. If you have more suggestions, feel free to comment.

File: sql/src/main/java/io/crate/executor/transport/NodeOperationTreeGenerator.java [hunk 10]
```diff
@@ -229,6 +230,30 @@ public Void visitMerge(Merge merge, NodeOperationTreeContext context) {
         return null;
     }
 
+    /**
+     * Generates the {@link NodeOperation}s for executing Union.
+     *
+     * We branch off for either side of the Union with a different input ids. In contrast
```
Comment (matriv) [inline sql/src/main/java/io/crate/executor/transport/NodeOperationTreeGenerator.java]: id instead of ids

File: sql/src/main/java/io/crate/planner/UnionPlan.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    public ResultDescription resultDescription() {
+        return this;
+    }
+
+    @Override
+    public void setDistributionInfo(DistributionInfo distributionInfo) {
+        mergePhase.distributionInfo(distributionInfo);
+    }
+
+    public MergePhase mergePhase() {
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/UnionPlan.java]: very Minor: I would move these 3 methods to the end to keep all the Overriden methods together

File: sql/src/main/java/io/crate/planner/operators/Union.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+            "union",
+            leftResultDesc.nodeIds().size() + rightResultDesc.nodeIds().size(),
+            2,
+            Collections.singletonList(plannerContext.handlerNode()),
+            leftResultDesc.streamOutputs(),
+            Collections.emptyList(),
+            DistributionInfo.DEFAULT_BROADCAST,
+            null
+        );
+
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/operators/Union.java]: extra empty line

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+    @Test
+    public void testUnionAllWithSystemTable() {
+        execute("select name from sys.nodes " +
+                "union all " +
+                "select text from t2");
+        int numResults = clusterService().state().nodes().getSize() + 3;
+        assertThat(response.rows().length, is(numResults));
+    }
+
```
Comment (matriv) [inline sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: extra empty line

**Comment by matriv** [comment]
@mxm  I would add some more Integ Tests:

-  System tables
-  "Scalar" subqueries  (return 1 value -> use MultiPhase plan)
- UnionAll as wrapped as a subquery, maybe also in a join or so.

File: sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java (hunk 9 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java]: Sorry guess I should've been more specific.

I wanted to know what's up with https://github.com/crate/crate/blob/b00f1e4d95e18c6a827084c86eacddec3aca044c/sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java#L133

because if you add the ObjectType.ID check like in this PR, the code below becomes effectively dead code as it will never be used. Which means the `ObjectColumnReference` is probably dead as well. 

Given that we always have to do source lookups for objects (if it's `obj` or `_doc['obj']` doesn't matter) , we could probably use the same expression implementation all the time.

**Comment by mfussenegger** [comment]
Not sure if you've seen my last comment since the change is shown as outdated:

https://github.com/crate/crate/pull/6535#discussion_r153014720

File: sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java (hunk 9 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java]: @mfussenegger I've replaced `ObjectColumnReference` and `GeoShapeColumnReference` with the doc source lookup. Seems to work fine. I wonder, dos this come at a performance cost because the lookup in the source is more expensive?

**Comment by mxm** [comment]
@mfussenegger Thx for pointing out. I've commented inline.

**Comment by mxm** [comment]
Thank you @matriv. I've addressed your comments.
